core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +120,146 @@ where
             hasher,
             hashes,
             binary_tree_size,
+            start_index: 0,
+            cache: vec![None; depth],
         }
     }
 
+    /// Returns `true` if the tree is empty.
+    pub fn is_empty(&self) -> bool {
+        self.start_index == 0 && self.hashes.is_empty()
+    }
+
     /// Returns the root hash of this tree.
     /// # Panics
-    /// Will panic if the constant below is invalid.
-    pub fn merkle_root(self) -> H256 {
+    /// Should never panic, unless there is a bug.
+    pub fn merkle_root(&self) -> H256 {
         if self.hashes.is_empty() {
             let depth = tree_depth_by_size(self.binary_tree_size);
-            self.hasher.empty_subtree_hash(depth)
-        } else {
-            self.compute_merkle_root_and_path(0, None)
+            if self.start_index == 0 {
+                return self.hasher.empty_subtree_hash(depth);
+            } else if self.start_index == self.binary_tree_size {
+                return self.cache[depth].expect("cache is invalid");
+            }
         }
+        self.compute_merkle_root_and_path(0, None, None)
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    /// # Panics
+    /// Panics if `index` is >= than the number of leaves in the tree.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        assert!(index < self.hashes.len(), "leaf index out of bounds");
+        let mut end_path = vec![];
+        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut end_path), None);
+        (
+            root_hash,
+            end_path.into_iter().map(Option::unwrap).collect(),
+        )
+    }
+
+    /// Returns the root hash and the Merkle proofs for a range of leafs.
+    /// The range is 0..length, where `0` is the leftmost untrimmed leaf.
+    /// # Panics
+    /// Panics if `length` is 0 or greater than the number of leaves in the tree.
+    pub fn merkle_root_and_paths_for_range(
+        &self,
+        length: usize,
+    ) -> (H256, Vec<Option<H256>>, Vec<Option<H256>>) {
+        assert!(length > 0, "range must not be empty");
+        assert!(length <= self.hashes.len(), "not enough leaves in the tree");
+        let mut right_path = vec![];
+        let root_hash =
+            self.compute_merkle_root_and_path(length - 1, Some(&mut right_path), Some(Side::Right));
+        (root_hash, self.cache.clone(), right_path)
+    }
+
+    /// Adds a raw hash to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push_hash(&mut self, leaf_hash: H256) {
```

Comment:
Sorry for not asking this previously: Is it intentional that adding hashes to the tree via this method and `from_hashes` is public? IMO, it's slightly confusing that there are multiple ways to add data to the tree.

Reply:
It is, since without them the only interface we have supports adding constant-sized arrays, whereas priority operations can be dynamically sized. 

Perhaps it would be better for `push` and `with_hasher` to accept slices instead?

Reply:
Probably yes, or make the tree parametric by the value type instead of using `const LEAF_SIZE` parameter (it doesn't makes sense for priority ops, does it?). Previously, it only accepted fixed-size arrays because we didn't need anything else 🙃 Just to clarify, what would be an empty hash for a priority op? The hash of an empty byte slice? You didn't refactor `HashEmptySubtree` or added more impls for it; and it feels arbitrary to use a hash of `[0_u8; 88]` for priority ops.

Reply:
This is a good point. I suggest we keep the `push_hash` and `from_hashes` since initializing the tree can be faster that way (since hashes are stored in Postgres, so no need to rehash when initializing). 

Re: `HashEmptySubtree` you are right, I'll need to have a separate impl for priority ops. I would rather come back to it in the next PR where I would integrate the tree into `eth_watch` and `eth_sender`. 

I like the idea of making the tree generic over leaf type, instead of over `LEAF_SIZE`, done in the last commit